test/cluster/catchup_test.go

```diff
@@ -0,0 +1,656 @@
+// - submit data txs.
+// - shutdown the leader.
+// - submit data txs.
+// - restart the server.
+// - make sure the server is in sync with previous txs.
+func TestLeaderRecoveryWithCatchup(t *testing.T) {
+	dir, err := ioutil.TempDir("", "int-test")
+	require.NoError(t, err)
+
+	nPort, pPort := getPorts(3)
+	setupConfig := &setup.Config{
+		NumberOfServers:     3,
+		TestDirAbsolutePath: dir,
+		BDBBinaryPath:       "../../bin/bdb",
+		CmdTimeout:          10 * time.Second,
+		BaseNodePort:        nPort,
+		BasePeerPort:        pPort,
+		CheckRedirectFunc: func(req *http.Request, via []*http.Request) error {
+			return errors.Errorf("Redirect blocked in test client: url: '%s', referrer: '%s', #via: %d", req.URL, req.Referer(), len(via))
+		},
+	}
+	c, err := setup.NewCluster(setupConfig)
+	require.NoError(t, err)
+	defer c.ShutdownAndCleanup()
+
+	require.NoError(t, c.Start())
+
+	leaderIndex := -1
+	require.Eventually(t, func() bool {
+		leaderIndex = c.AgreedLeader(t, 0, 1, 2)
+		return leaderIndex >= 0
+	}, 30*time.Second, 100*time.Millisecond)
+
+	//get current cluster config
+	configEnv, err := c.Servers[leaderIndex].QueryConfig(t)
+	require.NoError(t, err)
+	require.NotNil(t, configEnv)
+
+	newConfig := configEnv.GetResponse().GetConfig()
+	newConfig.ConsensusConfig.RaftConfig.SnapshotIntervalSize = 4 * 1024
+	version := configEnv.GetResponse().GetMetadata().GetVersion()
+	txsCount := 0
+
+	//change SnapshotIntervalSize to 4K
+	txID, rcpt, err := c.Servers[leaderIndex].SetConfigTx(t, newConfig, version)
+	require.NoError(t, err)
+	require.NotNil(t, rcpt)
+	require.True(t, txID != "")
+	require.True(t, len(rcpt.GetHeader().GetValidationInfo()) > 0)
+	require.Equal(t, types.Flag_VALID, rcpt.Header.ValidationInfo[rcpt.TxIndex].Flag)
+	txsCount++
+	t.Logf("tx submitted: %s, %+v", txID, rcpt)
+
+	keys := []string{"alice", "bob", "charlie", "dan"}
+	data := make([]byte, 1024)
+	for _, key := range keys {
+		txID, rcpt, err := c.Servers[leaderIndex].WriteDataTx(t, worldstate.DefaultDBName, key, data)
+		require.NoError(t, err)
+		require.NotNil(t, rcpt)
+		require.True(t, txID != "")
+		require.True(t, len(rcpt.GetHeader().GetValidationInfo()) > 0)
+		require.Equal(t, types.Flag_VALID, rcpt.Header.ValidationInfo[rcpt.TxIndex].Flag)
+		txsCount++
+		t.Logf("tx submitted: %s, %+v", txID, rcpt)
+	}
+
+	require.Eventually(t, func() bool {
+		return c.AgreedHeight(t, uint64(txsCount+1), 0, 1, 2)
+	}, 30*time.Second, 100*time.Millisecond)
+
+	//shutdown the leader
+	follower1 := (leaderIndex + 1) % 3
+	follower2 := (leaderIndex + 2) % 3
+	require.NoError(t, c.ShutdownServer(c.Servers[leaderIndex]))
+
+	//find new leader
+	newLeaderIndex := -1
+	require.Eventually(t, func() bool {
+		newLeaderIndex = c.AgreedLeader(t, follower1, follower2)
+		return newLeaderIndex >= 0
+	}, 30*time.Second, 100*time.Millisecond)
+
+	//submit txs with data size > SnapshotIntervalSize
+	keys = []string{"erin", "frank", "grace", "ivan"}
+	for _, key := range keys {
```

Comment:
I am thinking maybe it makes sense to write more than 4k, say 8k or 12k to create more than one snapshot?
Maybe overwrite each key several times?

Reply:
Done